libcudacxx/include/cuda/std/detail/libcxx/include/__bit_reference

```diff
@@ -22,1272 +20,1354 @@
+template <class _Cp, bool _IsConst>
+inline _CCCL_CONSTEXPR_CXX14 _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY __bit_iterator<_Cp, false>
+__copy_backward_aligned(
+  __bit_iterator<_Cp, _IsConst> __first, __bit_iterator<_Cp, _IsConst> __last, __bit_iterator<_Cp, false> __result)
 {
-    typedef __bit_iterator<_Cp, false> _It;
-    typedef typename _It::__storage_type __storage_type;
-    const int __bits_per_word = _It::__bits_per_word;
-    // do first partial word
-    if (__first.__ctz_ != 0)
+  using _In             = __bit_iterator<_Cp, _IsConst>;
+  using difference_type = typename _In::difference_type;
+  using __storage_type  = typename _In::__storage_type;
+
+  const int __bits_per_word = _In::__bits_per_word;
+  difference_type __n       = __last - __first;
+  if (__n > 0)
+  {
+    // do first word
+    if (__last.__ctz_ != 0)
     {
-        __storage_type __clz_f = static_cast<__storage_type>(__bits_per_word - __first.__ctz_);
-        __storage_type __dn = _CUDA_VSTD::min(__clz_f, __n);
-        __storage_type __m = (~__storage_type(0) << __first.__ctz_) & (~__storage_type(0) >> (__clz_f - __dn));
-        *__first.__seg_ &= ~__m;
-        __n -= __dn;
-        ++__first.__seg_;
+      difference_type __dn = _CUDA_VSTD::min(static_cast<difference_type>(__last.__ctz_), __n);
+      __n -= __dn;
+      unsigned __clz     = __bits_per_word - __last.__ctz_;
+      __storage_type __m = (~__storage_type(0) << (__last.__ctz_ - __dn)) & (~__storage_type(0) >> __clz);
+      __storage_type __b = *__last.__seg_ & __m;
+      *__result.__seg_ &= ~__m;
+      *__result.__seg_ |= __b;
+      __result.__ctz_ = static_cast<unsigned>(((-__dn & (__bits_per_word - 1)) + __result.__ctz_) % __bits_per_word);
+      // __last.__ctz_ = 0
     }
-    // do middle whole words
+    // __last.__ctz_ == 0 || __n == 0
+    // __result.__ctz_ == 0 || __n == 0
+    // do middle words
     __storage_type __nw = __n / __bits_per_word;
-    _CUDA_VSTD::memset(_CUDA_VSTD::__to_raw_pointer(__first.__seg_), 0, __nw * sizeof(__storage_type));
+    // NOTE(mdominiak): if the code in this block is always evaluated at constexpr, nvcc claims there's an invalid
+    // non-array pointer manipulation, so guard against that in constexpr
+    // nvcc's claim is correct, and technically we run into UB when __nw == 1 and __n == __bits_per_word,
+    // but since that's only pointer arithmetic UB, I don't think it's worth it to guard this at runtime
+    // on GCC <9, need to always do the check, because it doesn't have is_constant_evaluated
+#if defined(_CCCL_COMPILER_GCC) && _GNUC_VER < 900
+    if (__nw != 0 && (__nw != 1 || __n != __bits_per_word))
+#else
+    if (!__libcpp_is_constant_evaluated() || (__nw != 0 && (__nw != 1 || __n != __bits_per_word)))
+#endif
+    {
+      __result.__seg_ -= __nw;
+      __last.__seg_ -= __nw;
+      _CUDA_VSTD::copy_n(std::__to_address(__last.__seg_), __nw, std::__to_address(__result.__seg_));
```

Comment:
Critical: We should not use `std::` qualified functions.

Reply:
Ah yes, that's just me failing to add `/g` at the end of my sed here. Will fix.